src/memray/_memray/record_writer.cpp

```diff
@@ -93,23 +392,266 @@ RecordWriter::writeTrailer()
+            || !writeSimpleType(ContextSwitch{tid})
+            || !writeSimpleType(AggregatedRecordType::THREAD_RECORD)
+            || !writeString(thread_name.c_str()))
+        {
+            return false;
+        }
+    }
+
+    for (const auto& mappings : d_mappings_by_generation) {
+        if (!writeMappingsCommon(mappings)) {
+            return false;
+        }
+    }
+
+    UnresolvedNativeFrame last{};
+    for (const auto& record : d_native_frames) {
+        if (!writeSimpleType(AggregatedRecordType::NATIVE_TRACE_INDEX)
+            || !writeIntegralDelta(&last.ip, record.ip)
+            || !writeIntegralDelta(&last.index, record.index))
+        {
+            return false;
+        }
+    }
+
+    for (const auto& [frame_id, frame] : d_frames_by_id) {
+        if (!writeSimpleType(AggregatedRecordType::PYTHON_FRAME_INDEX) || !writeSimpleType(frame_id)
+            || !writeString(frame.function_name.c_str()) || !writeString(frame.filename.c_str())
+            || !writeSimpleType(frame.lineno) || !writeSimpleType(frame.is_entry_frame))
+        {
+            return false;
+        }
+    }
+
+    for (FrameTree::index_t index = d_python_frame_tree.minIndex();
+         index <= d_python_frame_tree.maxIndex();
+         ++index)
+    {
+        auto [frame_id, parent_index] = d_python_frame_tree.nextNode(index);
+
+        if (!writeSimpleType(AggregatedRecordType::PYTHON_TRACE_INDEX) || !writeSimpleType(frame_id)
+            || !writeSimpleType(parent_index))
+        {
+            return false;
+        }
+    }
+
+    d_high_water_mark_aggregator.visitAllocations([&](const AggregatedAllocation& allocation) {
+        if (allocation.n_allocations_in_high_water_mark == 0 && allocation.n_allocations_leaked == 0) {
+            return true;
+        }
+
+        return writeSimpleType(AggregatedRecordType::AGGREGATED_ALLOCATION)
+               && writeSimpleType(allocation);
+    });
+
+    // The FileSource will ignore trailing 0x00 bytes. This non-zero trailer
+    // marks the boundary between bytes we wrote and padding bytes.
+    if (!writeSimpleType(AggregatedRecordType::AGGREGATED_TRAILER)) {
+        return false;
+    }
+
+    return true;
 }
 
 std::unique_ptr<RecordWriter>
-RecordWriter::cloneInChildProcess()
+AggregatingRecordWriter::cloneInChildProcess()
 {
     std::unique_ptr<io::Sink> new_sink = d_sink->cloneInChildProcess();
     if (!new_sink) {
         return {};
     }
-    return std::make_unique<RecordWriter>(
+    return std::make_unique<AggregatingRecordWriter>(
             std::move(new_sink),
             d_header.command_line,
             d_header.native_traces);
 }
 
+bool
+AggregatingRecordWriter::writeRecord(const MemoryRecord& record)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    MemorySnapshot snapshot;
+    snapshot.ms_since_epoch = record.ms_since_epoch;
+    snapshot.rss = record.rss;
+    snapshot.heap = d_high_water_mark_aggregator.getCurrentHeapSize();
```

Comment:
This is a bit verbose. Any reason why we are not just doing:

```suggestion
    MemorySnapshot snapshot{record.ms_since_epoch, record.rss, d_high_water_mark_aggregator.getCurrentHeapSize());
```